Add support for octal and binary int literals

filbranden opened the pull request:
Simplify support for alternate bases using `int(..., base=0)` which auto-detects it using the standard Python syntax for numbers.

Octal numbers are useful to specify permission bits and umasks.

Binary numbers are not super useful... But considering we get them for free, let's allow them here too.

Add octal and binary number to a test case. Also update docs and add a snippet for release notes.

sarum9in commented on mesonbuild/mparser.py:
```diff
@@ -180,10 +179,7 @@ def lex(self, subdir):
-                    elif tid == 'hexnumber':
-                        tid = 'number'
-                        value = int(match_text, base=16)
+                        value = int(match_text, base=0)
```
Not sure if we care, but what makes me slightly concerned that we depend on python implementation here, e.g. if they add/change something meson syntax will be affected.

filbranden replied:
Yeah I can see that...

In particular, Python 2 used to recognize `0123` as octal syntax, the `0o123` was introduced in Python 2.7 and Python 3 no longer recognizes the `0123` syntax, only the `0o` prefix.

But given Meson syntax is Python-inspired, I think it's probably fine to keep this as is.

jpakkane commented:
What does `0123` resolve to? If it does not error out, should it?

jpakkane commented:
Then there should be a failing test for that.

This also has the potential for breaking backwards compatibility if there are projects out there using numbers with leading zeroes (unless that was already a hard error). Seems a bit unlikely, though.

filbranden commented:
> Then there should be a failing test for that.

Ok, I'll look into adding one... Obviously adding one to `test cases/common/68 number arithmetic/meson.build` is not an option, that's why I skipped it...

> This also has the potential for breaking backwards compatibility if there are projects out there using numbers with leading zeroes (unless that was already a hard error). Seems a bit unlikely, though.

Agreed. It would be possible to keep the backwards compatibility... But it would make the code more complex, and arguably the syntax would be confusing, since Python syntax itself would not allow that... So if you're fine with the small change of behavior, that's great.

I'll ping this PR once I've added a test for the breaking case of `0123`.

Cheers,
Filipe

jpakkane commented:
Failing tests are done by putting them in `unit tests/failing`. The test runner code takes care of the rest. Just make sure to manually check that it is failing due to the reason you expect it to rather than, say, typos in the build definition.

nirbheek commented:
> Failing tests are done by putting them in unit tests/failing. The test runner code takes care of the rest. Just make sure to manually check that it is failing due to the reason you expect it to rather than, say, typos in the build definition.

You can also add failing tests in `run_unittests.py` and then you can actually check that the failure was because of an expected condition (by matching the output with a regex)

filbranden commented:
Sorry, I got distracted for a bit...

I started to look at this one and I noticed that putting it under `unit tests/failing` doesn't work since it expects MesonException and this raises ValueError instead...

To keep that working the same way, I added this as a unit test and added an optional exception class argument to assertMesonRaises. I'm not too happy with that approach...

Today I was thinking of whether I should change the regex in mparser.py so the parser doesn't even recognize the number if it has a 0 prefix. The simplistic approach `[1-9]\d*` didn't work since `0` itself won't match it :-)

At this point, not sure what's the best way here... Use a `0|[1-9]\d*` regex in the parser definition?

Or in the `int(..., base=0)` conversion, catch ValueError and raise a MesonException in its place?

Or extend assertMesonRaises so it accepts a custom exception class for certain tests?

filbranden commented:
On second thought, I think the approach of `0|[1-9]\d*` is the most consistent one... Since for everything else, if something makes out of the parser, it should be handled... So I went with that.

I uploaded an updated patch set, including an extra commit with the additional "failing" test case, which now works as expected.

Cheers,
Filipe

nirbheek commented:
> I started to look at this one and I noticed that putting it under unit tests/failing doesn't work since it expects MesonException and this raises ValueError instead...

You should always catch the `ValueError` and output `MesonException` so that the user gets a nice error message instead of a traceback aka "crash".

filbranden commented:
> You should always catch the `ValueError` and output `MesonException` so that the user gets a nice error message instead of a traceback aka "crash".

I used a different approach (by tweaking the regex) but the end result should be the same...

As far as I can tell, `ValueError` should not be raised in this line of the code anymore, since the regex should only select tokens that are valid input for it. :-)

PTAL.

Cheers,
Filipe
